core/server/worker/src/main/java/alluxio/underfs/PagedUfsReader.java

```diff
@@ -244,36 +243,52 @@ public void close() throws IOException {
   private class UfsReadableChannel implements ReadableByteChannel {
     private final long mOffset;
     private volatile InputStream mUfsInStream;
-    private volatile ReadableByteChannel mUfsChannel;
     private volatile boolean mClosed = false;
 
     UfsReadableChannel(long offset) {
       mOffset = offset;
     }
 
     @Override
-    public int read(ByteBuffer dst) throws IOException {
+    public synchronized int read(ByteBuffer dst) throws IOException {
       if (mClosed) {
         throw new ClosedChannelException();
       }
+      int length = dst.remaining();
+      if (length == 0) {
+        return 0;
+      }
       String ufsPath = mUfsBlockOptions.getUfsPath();
       if (mUfsInStream == null) {
-        synchronized (this) {
-          if (mUfsInStream == null) {
-            try (CloseableResource<UnderFileSystem> ufsResource = mUfsClient.acquireUfsResource()) {
-              mUfsInStream = mUfsInStreamCache.acquire(
-                  ufsResource.get(),
-                  ufsPath,
-                  mFileId,
-                  OpenOptions.defaults()
-                      .setOffset(mUfsBlockOptions.getOffsetInFile() + mOffset)
-                      .setPositionShort(true));
-              mUfsChannel = Channels.newChannel(mUfsInStream);
-            }
-          }
+        try (CloseableResource<UnderFileSystem> ufsResource = mUfsClient.acquireUfsResource()) {
+          mUfsInStream = mUfsInStreamCache.acquire(
+              ufsResource.get(),
+              ufsPath,
+              mFileId,
+              OpenOptions.defaults()
+                  .setOffset(mUfsBlockOptions.getOffsetInFile() + mOffset)
+                  .setPositionShort(true));
         }
       }
-      return mUfsChannel.read(dst);
+      byte[] array;
+      int arrayOffset;
+      if (dst.hasArray()) {
+        array = dst.array();
+        arrayOffset = dst.arrayOffset();
+        int bytesRead = mUfsInStream.read(array, arrayOffset, length);
+        if (bytesRead > 0) {
+          dst.position(dst.position() + bytesRead);
+        }
+        return bytesRead;
+      } else {
+        array = new byte[length];
```

Comment:
Just had a discussion with @dbw9580 .
There are some possible optimizations.
1. Can we avoid allocating a new array each time for a request? We can allocate a buffer array in this class and use it for all read request? Since the read() is now synchronized, there should be no problem.
2. This class will be instantiated for each request? Can we make it a single long-loving object and use it for all request?
3. Can we avoid doing memory copying?


Maybe in most cases Worker will not use this code path? because the `dst.hasArray()` is true?

Reply:
The solution is to make UFS input streams accept a wider range of output buffer types. I think in @LuQQiu 's new PR we are implementing a UFS position reader which exactly does this.

Reply:
> Can we make it a single long-loving object and use it for all request

The plan is to make the readers thread safe, so reusing an object by setting a different starting offset on a second call will inevitably require synchronization.

To reduce memory pressure and also account for the inefficient use of the underlying resources caused by small (tiny) buffers, we should have *both* a positioned reader *and* a streaming reader.